src/libawkward/array/NumpyArray.cpp

```diff
@@ -4820,6 +4820,157 @@ namespace awkward {
                                         ptr_lib);
   }
 
+  const std::shared_ptr<NumpyArray> NumpyArray::cupy_copy_to(kernel::lib ptr_lib) const {
+    if (ptr_lib_ == ptr_lib) {
+      return std::make_shared<NumpyArray>(identities_,
+                                          parameters_,
+                                          ptr_,
+                                          shape_,
+                                          strides_,
+                                          byteoffset_,
+                                          itemsize_,
+                                          format_,
+                                          dtype_,
+                                          ptr_lib_);
+    }
+
+    ssize_t length = 1;
+    for (auto i : shape_) {
+      length = length * i;
+    }
+
+    std::shared_ptr<void> ptr;
+    Error err;
+
+    switch (dtype_) {
+      case util::dtype::boolean:
+        ptr = kernel::ptr_alloc<bool>(ptr_lib, (int64_t)length);
+        err = kernel::copy_to<bool>(
+          ptr_lib,
+          ptr_lib_,
+          reinterpret_cast<bool*>(ptr.get()),
+          reinterpret_cast<bool*>(data()),
+          length);
+        break;
+      case util::dtype::int8:
+        ptr = kernel::ptr_alloc<int8_t>(ptr_lib, (int64_t)length);
+        err = kernel::copy_to<int8_t>(
+          ptr_lib,
+          ptr_lib_,
+          reinterpret_cast<int8_t*>(ptr.get()),
+          reinterpret_cast<int8_t*>(data()),
+          length);
+        break;
```

Comment:
Maybe I should have brought this up earlier, but why does `copy_to` depend on data type? Without casting, the NumpyArray carries a `void*` (actually, `std::shared_ptr<void>`). To copy data from main memory to GPU or back, the CUDA copy-to-device or copy-to-host functions might take a `char*`, but that's one `reinterpret_cast`. You shouldn't need to specialize by dtype just to copy data (you have the `strides` and `shape` to determine the number of bytes to copy). That could save a lot of unnecessary functions.

Reply:
Yes, this makes sense. I could pass the `ptr` and the number of bytes to copy from the head of the ptr (`length() * itemsize()`) and that would be suffice for the `cudaMallocManaged`